docs/CompilerPerformance.md

```diff
@@ -0,0 +1,1142 @@
+
+Unfortunately judgements about when a file "needs recompiling" are themselves
+driven by an auxiliary data structure that summarizes the dependencies between
+files, and this data structure is necessarily a conservative approximation. The
+approximation is weaker than it should be, and as a result the driver often runs
+more frontend jobs than it should.
+
+#### Lazy resolution
+
+Swift source files contain names that refer to definitions outside the enclosing
+file, and frequently outside of the enclosing module. These "external"
+definitions are resolved lazily from two very different locations (both called
+"modules"):
+
+  - C/ObjC modules, provided by the Clang importer
+  - Serialized Swift modules
+
+Despite their differences, both kinds of modules support laziness in the Swift
+compiler in one crucial way: they are both kinds of _indexed_ binary file
+formats that permit loading _single definitions_ out of by name, without having
+to load the entire contents of the module.
+
+When the Swift compiler manages to be lazy and limit the number of definitions
+it tries to load from modules, it can be very fast; the file formats support
+very cheap access. But often the logic in the Swift compiler is unnecessarily
+conservative about exploiting this potential laziness, and so it loads more
+definitions than it should.
+
+### Summing up: high level picture of compilation performance
+
+Swift compilation performance varies _significantly_ by at least the following
+parameters:
+
+  - WMO vs. primary-file (non-WMO) mode
+  - Optimizing vs. non-optimizing mode
+  - Quantity of incremental work avoided (if in non-WMO)
+  - Quantity of external definitions lazily loaded
+
+When approaching Swift compilation performance, it's important to be aware of
+these parameters and keep them in mind, as they tend to frame the problem you're
+analyzing: changing one (or any of the factors influencing them, in a project)
+will likely completely change the resulting profile.
+
+## Known problem areas
+
+These are areas where we know the compiler has room for improvement,
+performance-wise, where it's worth searching for existing bugs on the topic,
+finding an existing team member who knows the area, and trying to relate the
+problem you're seeing to some of the existing strategies and plans for
+improvement:
+
+  - Incremental mode is over-approximate, runs too many subprocesses.
+  - Name resolution is over-eager, deserializes too many definitions.
+  - Too many referenced (non-primary-file) definitions are type-checked beyond
+    the point they need to be, during the quadratic phase.
+  - Expression type inference solves constraints inefficiently, and can
+    sometimes behave super-linearly or even exponentially.
+  - Periodically the analysis phase of a SIL optimization fails to cache
+    overlapping subproblems, causing a super-linear slowdown.
+  - Some SIL-to-IR lowerings (eg. large value types, ARC operations) can
```

Comment:
This is a bit misleading. We do generate more IR and there are parts of LLVM that do not handle this well, but it isn't due to ARC operations.

Reply:
Ok. I'll just reduce it to the single example. I'm not very familiar with the set of issues you're tracking in this area.